lib/datadog/profiling/component.rb

```diff
@@ -213,6 +213,17 @@ def self.build_profiler_component(settings:, agent_settings:, optional_tracer:)
           return false
         end
 
+        if RUBY_VERSION < '3.1'
+          Datadog.logger.warn(
+            "Current Ruby version (#{RUBY_VERSION}) supports forced object recycling which has a bug that the " \
+            'heap profiler is forced to work around to remain accurate. This workaround should be safe and ' \
+            'should even partially fix a runtime memory leak around object id mapping. Details about the workaround ' \
+            "can be found in https://github.com/DataDog/dd-trace-rb/pull/3360. If you're not comfortable with this, " \
+            'disable heap profiler or upgrade to Ruby >= 3.1 where forced object recycling was completely' \
+            'removed (https://bugs.ruby-lang.org/issues/18290).'
+          )
+        end
```

Comment:
I... don't quite like this warning as-is :thinking: 

Yes, there are unknown-unknowns in what we're doing, but I think anyone reading this warning would need to dive really, really, really deep to make an informed decision about this.

Furthermore, those customers that do decide "I am comfortable with this", will still see the giant scary warning every time, which is a bit meh.

On the other hand, I do see the relevance in documenting what we're doing in a user-visible way.

I'm thinking we could move documenting this to [the public docs](https://docs.datadoghq.com/profiler/profiler_troubleshooting/ruby/). Alternatively, we could change this to a debug, not warn-level log -- this would not show up by default, but if folks are investigating issues in ddtrace, they would get the message when they enable debug logging.

(I also have a few suggestions for the text itself, but I think the discussion on where we place it and when we show it may end up making some of the suggestions outdated so I'll hold those :wink: )

Reply:
Ack, sounds like a plan :+1: